### PR TITLE
refactoring-Level1 changes

### PR DESCRIPTION
**Pull request by alelom**
### Issues addressed by this PR
#86 
#89 

### Linked to Refactoring in BHoM_Engine
https://github.com/BHoM/BHoM_Engine/pull/1092

### Test files
Any script that does a Push - the more objects in the push, the better.

### Changelog
- Objects are now deep cloned at the start of the Push
- `ReplaceThroughApi()` has been simplified

### CODE REVIEW

**Comment by alelom** [comment]
CI tested :) 

However I meant to open this as a **draft PR** - forgot to do it and it Github doesn't allow conversion from review->draft.

Closing and reopening as draft PR. I'll do the change in BHoM_Engine first.

**Comment by alelom** [comment]
Something strange:
image
needs investigation

**Comment by alelom** [comment]
Raised [issue in BHoM_Engine](https://github.com/BHoM/BHoM_Engine/issues/1088)
As explained in https://github.com/BHoM/BHoM_Engine/pull/1089
removing a reference to a Nuget package now solves this issue.

**Comment by FraserGreenroyd** [comment]
@IsakNaslundBh to review and feedback to @alelom who will then finish up when feedback is received.

**Comment by adecler** [comment]
Overall, looks like a good first step. 
The code will probably need to be rearranged to make it easier to follow what is going on without having to rely on the diagrams you drew. But that's mainly cosmetic and the code looks sound on a functional level.

**Comment by IsakNaslundBh** [comment]
Have had a first look through the changes, and to me they make sense. 👍 

Will switch to this branch and run on it for some time and check that things still behave as they should.

Will also try to find some time to do some more focused testing, specifically on GSA which have the full tag functionality implemented.

File: BHoM_Adapter/CRUD/Replace.cs [hunk 1]
```diff
@@ -160,40 +188,35 @@ public abstract partial class BHoMAdapter
...
 
-            // Check and map properties
-            diagram2.Intersection.ForEach(x => x.Item1.MapSpecialProperties(x.Item2, AdapterId));
+            // Delete exclusive existing objects that do not have any other tag except the current tag from the model
+            Delete(typeof(T), existingObjs_exclusive.Where(x => x.Tags.Count == 0).Select(x => x.CustomData[AdapterId]));
 
-            //Update the tags
-            UpdateProperty(typeof(T), diagram2.OnlySet2.Select(x => x.CustomData[AdapterId]), "Tags", diagram2.OnlySet2.Select(x => x.Tags));
+            //Update the tags for the rest of the existing objects in the model
+            UpdateProperty(typeof(T), existingObjs_exclusive.Select(x => x.CustomData[AdapterId]), "Tags", existingObjs_exclusive.Select(x => x.Tags));
```
Comment (IsakNaslundBh) [inline BHoM_Adapter/CRUD/Replace.cs]: Think this only should be:

`existingObjs_exclusive.Where(x => x.Tags.Count != 0)`

that need to get the tags updated?

Comment (alelom) [reply BHoM_Adapter/CRUD/Replace.cs]: That is true! Excellently spotted! Thanks!

**Comment by FraserGreenroyd** [comment]
@alelom can you remove `BHoM_Adapter` from the PR title ready for when this is merged :smile:

**Comment by adecler** [comment]
The new changes look fine to me but I would really prefer to have a review from @IsakNaslundBh

**Comment by IsakNaslundBh** [comment]
Have been running on this branch for a week or two now with no problem. Will make one more focused review first thing tomorrow morning. Sorry this has dragged out this long @alelom

**Pull request merged**